ADR for Self-Description (#35)
agera-edc/IdentityHub: docs/developer/decision-records/2022-07-29-self-description/README.md

# Identity Hub - Self Description

This document explains how the Self-Description of a dataspace participant will be stored into the Identity Hub and retrieved from it.

## Context

Gaia-X defines a set of rules that define the minimum baseline to be part of a Gaia-X Ecosystem. This set of rules is the so-called
[Gaia-X Trust Framework](https://gaia-x.gitlab.io/policy-rules-committee/trust-framework/), which is centered around the _Self-Description_ document.

Each dataspace participant (consumer, provider, federator) is represented by a Self-Description document.

According to [spec](https://gaia-x.eu/wp-content/uploads/2022/06/Gaia-x-Architecture-Document-22.04-Release.pdf),
"Self-Descriptions are W3C Verifiable Presentations in the JSON-LD format. Self-Description consist of a list of
Verifiable Credentials. Verifiable Credentials themselves contain a list of Claims: assertions about Entities

> expressed in the RDF data model. Both Verifiable Credentials and Verifiable Presentations come with

Reviewer: nit: a link to an RDF document or spec would be useful

cryptographic signatures to increase the level of trust".

Note that a Self-Description can aggregate Verifiable Credentials from different issuers, each having its own cryptographic signature.

> An example of valid Self-Description can be found [here](https://gitlab.com/gaia-x/gaia-x-community/gx-hackathon/gx-hackathon-4/-/blob/main/Example%20Self-Descriptions/participantA1digital.json), which corresponds to the one used during the Gaia-X Hackathon #4.

Reviewer 2: use a permalink?

> Long term, the Verifiable Credentials from the Self-Description should be stored into the Identity Hub so that they can latter

Reviewer 3: The reason to serve the self-description document via the IdentityHub ist to be able to dynamically generate the self-description at a later stage. If we serve the self-description already via the IdentityHub now, nothing changes for consumers accessing self-descriptions at a later stage when self-descriptions are generated.

We could add this information to the ADR.

been verified and used for the policy enforcement (see [here](../2022-07-01-get-claims/README.md) for more details).

However, as specifications of the Self-Description documents are not definitive yet, it is for now preferred to restrict the usage of the
Self-Description document for display only. More specifically, we aim at exposing the Self-Description document through a dedicated REST endpoint,
which is resolved from the DID document of the participant.

## Assumptions

- Self-Description document is static, i.e. no update of the Self-Description is supported in this version

> - Self-Description document is provided at startup , i.e. no need for an endpoint allowing to update the Self-Description.

Reviewer 4: nit: maybe we can also add that in this version we _do not_ expect that this Self-Description document can be verified by Gaia-X.

## Implementation proposal

### Self-Description storage

> The Self-Description JSON document must be stored in the Identity Hub so that it can latter been served by the endpoint.

Reviewer 3: ```suggestion
The Self-Description JSON document must be stored in the Identity Hub so that it can later been served by the endpoint.
```

A tentative interface for the `SelfDescriptionStore` would be:

```java

> public class SelfDescriptionStore {

Reviewer: Do we really need a store? Or is a reference to a static file enough for the first step?

    void store(Object selfDescription);

    Object get();
}
```

An in-memory implementation of this store will also be provided.

> The populated at startup of the Identity Hub by reading the Self-Description from a static JSON file.

Reviewer 3: ```suggestion
The store is populated at startup of the Identity Hub by reading the Self-Description from a static JSON file.
```

### Self-Description endpoint

> The Self-Description can be retrieved from the store through a REST `GET` endpoint under the path `identity-hub/self-description`.

Reviewer 4: nit:
```suggestion
The Self-Description can be retrieved from identity hub through a REST `GET` endpoint under the well-known path `/self-description`.
```
